io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,592 @@
+         if (!subdir) {
+            subdir = (TDirectory *)key->ReadObj();
+         }
+         TMPIFile::R__DeleteObject(subdir, withReset);
+      } else {
+         Bool_t todelete = kFALSE;
+         if (withReset) {
+            todelete = (0 != cl->GetResetAfterMerge());
+         } else {
+            todelete = (0 == cl->GetResetAfterMerge());
+         }
+         if (todelete) {
+            key->Delete();
+            dir->GetListOfKeys()->Remove(key);
+            delete key;
+         }
+      }
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// return True if this is the Collector rank, otherwise False
+
+Bool_t TMPIFile::IsCollector()
+{
+   if (this->fMPILocalRank) {
+      return kFALSE;
+   }
+   return kTRUE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Called by the Workers only: Copies the current content in memory and
+/// sends it asynchronously to the Collector for merging and writing to disk.
+///
+
+void TMPIFile::CreateBufferAndSend()
+{
+   if (this->IsCollector()) {
+      SysError("CreateBufferAndSend", " should not be called by a collector");
+      exit(1);
+   }
+   this->Write();
+   Int_t count = this->GetEND();
+   fSendBuf = new char[count];
+   this->CopyTo(fSendBuf, count);
+   MPI_Isend(fSendBuf, count, MPI_CHAR, 0, fMPIColor, fSubComm, &fMPIRequest);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// For Workers: Creates an empty buffer and sends it to the Collector.
+/// This indicates the completion of the worker.
+
+void TMPIFile::CreateEmptyBufferAndSend()
+{
+   if (this->IsCollector()) {
+      return;
+   }
+
+   if (!IsReceived()) {
+      MPI_Wait(&fMPIRequest, MPI_STATUS_IGNORE);
+      delete[] fSendBuf;
+   }
+   Int_t sent = MPI_Send(fSendBuf, 0, MPI_CHAR, 0, fMPIColor, fSubComm);
+   if (sent) {
+      delete[] fSendBuf;
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Called by the Workers only: Called periodically by workers and triggers
+/// the sending of data to the Collector for writing.
+
+void TMPIFile::Sync()
+{
+   // check if the previous send request is accepted by master.
+   if (IsReceived()) { // if accpeted create and send current batch
+      CreateBufferAndSend();
+   } else {
+      // if not accepted wait until received by master
+      MPI_Wait(&fMPIRequest, MPI_STATUS_IGNORE);
+      delete[] fSendBuf; // empty the buffer once received by master
+      CreateBufferAndSend();
+   }
+   this->ResetAfterMerge((TFileMergeInfo *)0);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Closes the file. For Worker ranks, this function will signal to the
+/// Collector that the Worker has exited. It also closes the inherited TMemFile.
+
+void TMPIFile::Close(Option_t *option)
+{
+   // sends empty buffer
+   CreateEmptyBufferAndSend();
```

Comment:
What happens if one calls `Close()` on an already closed `TMPIFile`? It could be the case a user calls it, then the destructor calls it again.

Reply:
In the destructor, Close() is called only if IsOpen() of the file is true. I will add if IsOpen() at the very beginning of Close() to make sure it's safe to call it multiple times by accident.

Reply:
and then you can make the call is the destructor unconditional. (i.e. no need to check twice)

Reply:
done